metron-platform/metron-enrichment/Performance.md

```diff
@@ -0,0 +1,522 @@
+
+#### `metron.threadpool.type`
+
+The enrichment bolt maintains a static thread pool that is used to execute each enrichment.  This thread pool is shared by all of the executors running within the same worker.  
+
+Defines the type of thread pool used.  This value can be either "FIXED" or "WORK_STEALING".
+
+Currently, this value must be manually defined within the flux file at `$METRON_HOME/flux/enrichment/remote-unified.yaml`.  This value cannot be altered within Ambari.
+
+### Benchmark Results
+
+This section describes one execution of these benchmarks to help provide an understanding of what reasonably tuned parameters might look-like.  
+
+These parameters and the throughput reported are highly dependent on the workload and resources available. The throughput is what was achievable given a reasonable amount of tuning on a small, dedicated cluster.  The throughput is largely dependent on the enrichments performed and the distribution of data within the incoming telemetry.
+
+The Enrichment topology has been show to scale relatively linearly.  Adding more resources allows for more complex enrichments, across more diverse data sets, at higher volumes.  The throughput that one might see in production largely depends on how much hardware can be committed to the task.  
+
+#### Environment
+
+* Apache Metron 0.4.3 (pre-release) March, 2018
+	* This included [a patch to the underlying caching mechanism](https://github.com/apache/metron/pull/947) that greatly improves performance.
+
+* Cisco UCS nodes
+ 	* 32 core, 64-bit CPU (Intel(R) Xeon(R) CPU E5-2630 v3 @ 2.40GHz)
+	* 256 GB RAM
+	* x2 10G NIC bonded
+	* x4 6TB 7200 RPM disks
+	
+* Storm Supervisors are isolated and running on a dedicated set of 3 nodes.
+
+* Kafka Brokers are isolated and running on a separate, dedicated set of 3 nodes.
+
+#### Results
+
+* These benchmarks executed all 3 enrichments simultaneously; the [Geo IP Enrichment](#geo-ip-enrichment), [Stellar Enrichment](#stellar-enrichment) and the [HBase Enrichment](#hbase-enrichment).
+
+* The data used to drive the benchmark includes 10,000 unique IP addresses.  The telemetry was populated with IP addresses such that 10% of these IPs were chosen 80% of the time.  This bias was designed to mimic the typical distribution seen in real-world telemetry.
+
+* The Unified Enrichment topology was able to sustain 308,000 events per second on a small, dedicated 3 node cluster.
+
+* The values used to achieve these results with the Unified Enrichment topology follows.
+	```
+	enrichment.workers=9
+	enrichment.acker.executors=9
+	enrichment.join.cache.size=100000
+	kafka.spout.parallelism=27
+	enrichment.join.parallelism=54
+	threat.intel.join.parallelism=9
+	kafka.writer.parallelism=27
+	topology.worker.childopts=-XX:+UseG1GC -Xms8g -Xmx8g -XX:MaxGCPauseMillis=100
```

Comment:
It may be worth while to mention that we chose to use G1GC here because of the small object allocation issues with the cache made us concerned about GC issues (hence the max GC pause cap).  A small blurb about "If you have a very large cache and are experiencing performance bottlenecking in the enrichment bolts,  you might turn on GC logging" might be in order, just to point out that it's an area that we did look at.

Reply:
Makes sense..  I addressed this in the latest commit.  Let me know if that works for you.